### ISSUE TITLE
Dynmap web shared on other server, render/json file writing lagging/crashing server?

### ISSUE THREAD

**Issue by LigerXT5** [webbukkit/dynmap#2775]
Excuse my failed copy/paste of the log, while placing it in a code bracket, resulting in deleting what I had pretyped...

I've been having an odd issue with Dynmap locking up, and sometimes crashing the server. Either when doing a full render, or not. When other players are on or not.

Instead of hosting the Web folder locally, due to storage limitations, I have the web folder symlinked to another server via FTP. I'm thinking it's my host (both webhost and vps, same host).

The issue seems to be mostly random. I can go multiple days without issues, and some days have issues. Again, seems to happen when there are players or not, full render or not. Though, more frequently with full render, but can still go a day or so with issues some times.

[17:13:27] [Dynmap Render Thread/INFO]: [dynmap] Full render of map 'surface' of 'KirFam' in progress - 660000 tiles rendered (552.01 msec/tile, 542.38 msec per render)
[17:21:46] [Dynmap Render Thread/INFO]: [dynmap] Full render of map 'surface' of 'KirFam' in progress - 661000 tiles rendered (551.92 msec/tile, 542.21 msec per render)
[17:30:08] [Dynmap Render Thread/INFO]: [dynmap] Full render of map 'surface' of 'KirFam' in progress - 662000 tiles rendered (551.84 msec/tile, 542.04 msec per render)
[17:36:12] [Paper Watchdog Thread/ERROR]: --- DO NOT REPORT THIS TO PAPER - THIS IS NOT A BUG OR A CRASH  - git-Paper-77 (MC: 1.15.2) ---
[17:36:12] [Paper Watchdog Thread/ERROR]: The server has not responded for 10 seconds! Creating thread dump
[17:36:12] [Paper Watchdog Thread/ERROR]: ------------------------------
[17:36:12] [Paper Watchdog Thread/ERROR]: Server thread dump (Look for plugins here before reporting to Paper!):
[17:36:13] [Paper Watchdog Thread/ERROR]: ------------------------------
[17:36:13] [Paper Watchdog Thread/ERROR]: Current Thread: Server thread
[17:36:13] [Paper Watchdog Thread/ERROR]: 	PID: 20 | Suspended: false | Native: true | State: RUNNABLE
[17:36:13] [Paper Watchdog Thread/ERROR]: 	Stack:
[17:36:13] [Paper Watchdog Thread/ERROR]: 		java.base@11.0.5/java.io.UnixFileSystem.getBooleanAttributes0(Native Method)
[17:36:13] [Paper Watchdog Thread/ERROR]: 		java.base@11.0.5/java.io.UnixFileSystem.getBooleanAttributes(UnixFileSystem.java:248)
[17:36:13] [Paper Watchdog Thread/ERROR]: 		java.base@11.0.5/java.io.File.exists(File.java:820)
[17:36:13] [Paper Watchdog Thread/ERROR]: 		org.dynmap.storage.MapStorage.getStandaloneFile(MapStorage.java:320)
[17:36:13] [Paper Watchdog Thread/ERROR]: 		org.dynmap.JsonFileClientUpdateComponent.handleWebChat(JsonFileClientUpdateComponent.java:373)
[17:36:13] [Paper Watchdog Thread/ERROR]: 		org.dynmap.JsonFileClientUpdateComponent$1.run(JsonFileClientUpdateComponent.java:152)
[17:36:13] [Paper Watchdog Thread/ERROR]: 		app//org.bukkit.craftbukkit.v1_15_R1.scheduler.CraftTask.run(CraftTask.java:84)
[17:36:13] [Paper Watchdog Thread/ERROR]: 		app//org.bukkit.craftbukkit.v1_15_R1.scheduler.CraftScheduler.mainThreadHeartbeat(CraftScheduler.java:452)
[17:36:13] [Paper Watchdog Thread/ERROR]: 		app//net.minecraft.server.v1_15_R1.MinecraftServer.b(MinecraftServer.java:1185)
[17:36:13] [Paper Watchdog Thread/ERROR]: 		app//net.minecraft.server.v1_15_R1.DedicatedServer.b(DedicatedServer.java:423)
[17:36:13] [Paper Watchdog Thread/ERROR]: 		app//net.minecraft.server.v1_15_R1.MinecraftServer.a(MinecraftServer.java:1112)
[17:36:13] [Paper Watchdog Thread/ERROR]: 		app//net.minecraft.server.v1_15_R1.MinecraftServer.run(MinecraftServer.java:934)
[17:36:13] [Paper Watchdog Thread/ERROR]: 		java.base@11.0.5/java.lang.Thread.run(Thread.java:834)
[17:36:13] [Paper Watchdog Thread/ERROR]: ------------------------------
[17:36:13] [Paper Watchdog Thread/ERROR]: --- DO NOT REPORT THIS TO PAPER - THIS IS NOT A BUG OR A CRASH ---
[17:36:13] [Paper Watchdog Thread/ERROR]: ------------------------------
[17:36:17] [Paper Watchdog Thread/ERROR]: --- DO NOT REPORT THIS TO PAPER - THIS IS NOT A BUG OR A CRASH  - git-Paper-77 (MC: 1.15.2) ---
[17:36:17] [Paper Watchdog Thread/ERROR]: The server has not responded for 15 seconds! Creating thread dump
[17:36:17] [Paper Watchdog Thread/ERROR]: ------------------------------
[17:36:17] [Paper Watchdog Thread/ERROR]: Server thread dump (Look for plugins here before reporting to Paper!):
[17:36:17] [Paper Watchdog Thread/ERROR]: ------------------------------
[17:36:17] [Paper Watchdog Thread/ERROR]: Current Thread: Server thread
[17:36:17] [Paper Watchdog Thread/ERROR]: 	PID: 20 | Suspended: false | Native: true | State: RUNNABLE
[17:36:17] [Paper Watchdog Thread/ERROR]: 	Stack:
[17:36:17] [Paper Watchdog Thread/ERROR]: 		java.base@11.0.5/java.io.UnixFileSystem.getBooleanAttributes0(Native Method)
[17:36:17] [Paper Watchdog Thread/ERROR]: 		java.base@11.0.5/java.io.UnixFileSystem.getBooleanAttributes(UnixFileSystem.java:248)
[17:36:17] [Paper Watchdog Thread/ERROR]: 		java.base@11.0.5/java.io.File.exists(File.java:820)
[17:36:17] [Paper Watchdog Thread/ERROR]: 		org.dynmap.storage.MapStorage.getStandaloneFile(MapStorage.java:320)
[17:36:17] [Paper Watchdog Thread/ERROR]: 		org.dynmap.JsonFileClientUpdateComponent.handleWebChat(JsonFileClientUpdateComponent.java:373)
[17:36:17] [Paper Watchdog Thread/ERROR]: 		org.dynmap.JsonFileClientUpdateComponent$1.run(JsonFileClientUpdateComponent.java:152)
[17:36:17] [Paper Watchdog Thread/ERROR]: 		app//org.bukkit.craftbukkit.v1_15_R1.scheduler.CraftTask.run(CraftTask.java:84)
[17:36:17] [Paper Watchdog Thread/ERROR]: 		app//org.bukkit.craftbukkit.v1_15_R1.scheduler.CraftScheduler.mainThreadHeartbeat(CraftScheduler.java:452)
[17:36:17] [Paper Watchdog Thread/ERROR]: 		app//net.minecraft.server.v1_15_R1.MinecraftServer.b(MinecraftServer.java:1185)
[17:36:17] [Paper Watchdog Thread/ERROR]: 		app//net.minecraft.server.v1_15_R1.DedicatedServer.b(DedicatedServer.java:423)
[17:36:17] [Paper Watchdog Thread/ERROR]: 		app//net.minecraft.server.v1_15_R1.MinecraftServer.a(MinecraftServer.java:1112)
[17:36:17] [Paper Watchdog Thread/ERROR]: 		app//net.minecraft.server.v1_15_R1.MinecraftServer.run(MinecraftServer.java:934)
[17:36:17] [Paper Watchdog Thread/ERROR]: 		java.base@11.0.5/java.lang.Thread.run(Thread.java:834)
[17:36:17] [Paper Watchdog Thread/ERROR]: ------------------------------
[17:36:17] [Paper Watchdog Thread/ERROR]: --- DO NOT REPORT THIS TO PAPER - THIS IS NOT A BUG OR A CRASH ---
[17:36:17] [Paper Watchdog Thread/ERROR]: ------------------------------

**Comment by baconwaifu** [comment]
the watchdog firing is indeed caused by hung file IO. Dynmap supports remote storage natively via the database backends; just specify a remote database address. 
NOTE: if you want webchat with a standalone webserver, you need MySQL or MariaDB; the PostgreSQL backend doesn't support standalone mode in the current release.

The component that's hung, webchat, is not very well implemented at the moment to begin with. It *works*, but could be done significantly better with modern APIs such as websockets.

**Comment by LigerXT5** [comment]
Thanks for the info, it helps my understanding of what's flagging the issue.

I've tried using a database before, MySQL, however the database my host gives with the webserver, shot that idea down, as the database became very big. I used the FTP Symlink, as the storage for web hosting is "unlimited" and I haven't received any complaints.

I would store on the local VPS, however the storage itself is limited, 400GB, and I'm running a small handful of small MC servers, currently with just one with Dynmap (potentially two or three later one, but limited sized MC servers).

Thanks for the update. I'll see what work arounds I can find, possibly something like a caching/buffer system for when the FTP Symlink acts up, disconnect, reconnects, or what it is doing to cause issues.
